File: 4/ans.pl
input(Xs) :-
  read_stream_to_codes(current_input, LineCodes),
  string_codes(Line, LineCodes),
  split_string(Line, '\n ', '', Entries),
  entry_to_kv(Entries, KVs),
  split_by_empty(KVs, Xs).

entry_to_kv([], []).
entry_to_kv([X|Xs], [KV|Rest]) :-
  split_string(X, ':', '', KV),
  entry_to_kv(Xs, Rest).

split_by_empty([], []).
split_by_empty(Input, [Ans|Rest]) :-
  append([Ans, [[""]], NewInput], Input),
  split_by_empty(NewInput, Rest).

keys([], []).
keys([[K,_V]|Xs], [K|Rest]) :- keys(Xs, Rest).

part1([], 0).
part1([X|Xs], Ans) :-
  part1(Xs, NewAns),
  (validate_fields_existence(X) -> Ans is NewAns + 1; Ans is NewAns).

part2([], 0).
part2([X|Xs], Ans) :-
  part2(Xs, NewAns),
  (validate(X) -> Ans is NewAns + 1 ; Ans is NewAns).

validate(X) :- validate_fields_content(X), validate_fields_existence(X).

validate_fields_existence(X) :-
  keys(X, Keys),
  Required = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"],
  intersection(Keys, Required, Intersection),
  length(Required, RequiredLength),
  length(Intersection, IntersectionLength),
  RequiredLength =:= IntersectionLength.

validate_fields_content([]).
validate_fields_content([["byr", BirthYearString]|Xs]) :-
  string_length(BirthYearString, 4),
  number_string(BirthYear, BirthYearString),
  BirthYear >= 1920,
  BirthYear =< 2002,
  validate_fields_content(Xs).
validate_fields_content([["iyr", IssueYearString]|Xs]) :-
  string_length(IssueYearString, 4),
  number_string(IssueYear, IssueYearString),
  IssueYear >= 2010,
  IssueYear =< 2020,
  validate_fields_content(Xs).
validate_fields_content([["eyr", ExpirationYearString]|Xs]) :-
  string_length(ExpirationYearString, 4),
  number_string(ExpirationYear, ExpirationYearString),
  ExpirationYear >= 2020,
  ExpirationYear =< 2030,
  validate_fields_content(Xs).
validate_fields_content([["hgt", HeightString]|Xs]) :-
  sub_string(HeightString, _, 2, 0, Unit),
  sub_string(HeightString, 0, _, 2, HeightWithoutUnitString),
  number_string(HeightWithoutUnit, HeightWithoutUnitString),
  validate_height(HeightWithoutUnit, Unit),
  validate_fields_content(Xs).
validate_fields_content([["hcl", HairColorString]|Xs]) :-
  string_concat("#", Rest, HairColorString),
  string_codes(Rest, RestCodes),
  validate_hair(RestCodes),
  validate_fields_content(Xs).
validate_fields_content([["ecl", EyeColor]|Xs]) :-
  member(EyeColor, ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]),
  validate_fields_content(Xs).
validate_fields_content([["pid", PassportId]|Xs]) :-
  string_length(PassportId, 9),
  number_string(_, PassportId),
  validate_fields_content(Xs).
validate_fields_content([["cid", _]|Xs]) :- validate_fields_content(Xs).

validate_height(Height, "cm") :- Height >= 150, Height =< 193.
validate_height(Height, "in") :- Height >= 59, Height =< 76.

validate_hair([]).
validate_hair([X|Xs]) :-
  ((X >= 48, X =< 57); (X >= 97, X =< 102)),
  validate_hair(Xs).



main :-
  input(Xs),
  part1(Xs, Ans1),
  writeln(Ans1),
  part2(Xs, Ans2),
  writeln(Ans2).

File: 04/ans.zig
const std = @import("std");
const stdin = std.io.getStdIn().reader();
const stdout = std.io.getStdOut().writer();

const ArrayList = std.ArrayList;
const HashMap = std.StringHashMap([]const u8);

const REQUIRED_FIELDS = [_][]const u8{ "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid" };
const VALID_EYE_COLORS = [_][]const u8{ "amb", "blu", "brn", "gry", "grn", "hzl", "oth" };

fn readInput(allocator: *std.mem.Allocator) !ArrayList(HashMap) {
    var list = ArrayList(HashMap).init(allocator);
    const input = try stdin.readAllAlloc(allocator, 16777216);
    var lineIterator = std.mem.split(input, "\n");
    var curList = HashMap.init(allocator);
    while (lineIterator.next()) |line| {
        if (line.len == 0) {
            try list.append(curList);
            curList = HashMap.init(allocator);
            continue;
        }
        var entryIterator = std.mem.split(line, " ");
        while (entryIterator.next()) |item| {
            var kvIterator = std.mem.split(item, ":");
            const k = kvIterator.next().?;
            const v = kvIterator.next().?;
            try curList.putNoClobber(k, v);
        }
    }
    return list;
}

fn part1(input: ArrayList(HashMap)) !void {
    var valid: i32 = 0;
    outer: for (input.items) |passport| {
        for (REQUIRED_FIELDS) |field| {
            if (!passport.contains(field)) {
                continue :outer;
            }
        }
        valid += 1;
    }
    try stdout.print("{}\n", .{valid});
}

fn part2(input: ArrayList(HashMap)) !void {
    var valid: i32 = 0;
    outer: for (input.items) |passport| {
        for (REQUIRED_FIELDS) |field| {
            if (!passport.contains(field)) {
                continue :outer;
            }
        }
        if (!(try valid_passport(passport))) {
            continue;
        }
        valid += 1;
    }
    try stdout.print("{}\n", .{valid});
}

fn valid_passport(passport: HashMap) !bool {
    const birthYearString = passport.get("byr").?;
    if (birthYearString.len != 4) {
        return false;
    }
    const birthYear = try std.fmt.parseUnsigned(u32, birthYearString, 10);
    if (!(birthYear >= 1920 and birthYear <= 2002)) {
        return false;
    }

    const issueYearString = passport.get("iyr").?;
    if (issueYearString.len != 4) {
        return false;
    }
    const issueYear = try std.fmt.parseUnsigned(u32, issueYearString, 10);
    if (!(issueYear >= 2010 and issueYear <= 2020)) {
        return false;
    }

    const expirationYearString = passport.get("eyr").?;
    if (expirationYearString.len != 4) {
        return false;
    }
    const expirationYear = try std.fmt.parseUnsigned(u32, expirationYearString, 10);
    if (!(expirationYear >= 2020 and expirationYear <= 2030)) {
        return false;
    }

    const heightString = passport.get("hgt").?;
    const unit = heightString[heightString.len - 2 ..];
    const height = try std.fmt.parseUnsigned(u32, heightString[0 .. heightString.len - 2], 10);
    if (std.mem.eql(u8, unit, "cm")) {
        if (!(height >= 150 and height <= 193)) {
            return false;
        }
    } else if (std.mem.eql(u8, unit, "in")) {
        if (!(height >= 59 and height <= 76)) {
            return false;
        }
    } else {
        return false;
    }

    const hairColorString = passport.get("hcl").?;
    if (hairColorString.len != 7) {
        return false;
    }
    if (hairColorString[0] != '#') {
        return false;
    }
    for (hairColorString[1..6]) |c| {
        if (!((c >= '0' and c <= '9') or (c >= 'a' and c <= 'f'))) {
            return false;
        }
    }

    const eyeColorString = passport.get("ecl").?;
    if (!validEyeColor(eyeColorString)) {
        return false;
    }

    const passportId = passport.get("pid").?;
    if (passportId.len != 9) {
        return false;
    }
    _ = std.fmt.parseUnsigned(u32, passportId, 10) catch |_| return false;

    return true;
}

fn validEyeColor(eyeColorString: []const u8) bool {
    for (VALID_EYE_COLORS) |color| {
        if (std.mem.eql(u8, color, eyeColorString)) {
            return true;
        }
    }
    return false;
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;

    const input = try readInput(allocator);
    try part1(input);
    try part2(input);
}
